src/forms/inputs/file-input/file-input.js

```diff
@@ -49,141 +50,133 @@ import { noop, generateInputErrorId, removeAt } from '../../../utils'
  */
 
 const propTypes = {
-  ...fieldPropTypes,
-  onLoad: PropTypes.func,
+  ...fileInputPropTypes,
+  onRead: PropTypes.func,
   thumbnail: PropTypes.string,
   hidePreview: PropTypes.bool,
   className: PropTypes.string,
   previewComponent: PropTypes.func,
   children: PropTypes.node,
   multiple: PropTypes.bool,
   onRemove: PropTypes.func,
-  removeText: PropTypes.string,
+  selectText: PropTypes.string,
 }
 
 const defaultProps = {
   multiple: false,
-  onLoad: noop,
+  onRead: identity,
   onRemove: noop,
-}
-
-function readFile (file) {
-  return new Promise((resolve, reject) => {
-    const reader = new FileReader()
-    reader.onload = (readEvent) => {
-      resolve(readEvent.target.result)
-    }
-    reader.onerror = reject
-    
-    reader.readAsDataURL(file)
-  })
+  selectText: '',
 }
 
 class FileInput extends React.Component {
-
   constructor (props) {
     super(props)
-    this.state = { files: [] }
-    this.loadFiles = this.loadFiles.bind(this)
+    
+    this.state = { errors: null }
     this.onChange = this.onChange.bind(this)
     this.removeFile = this.removeFile.bind(this)
-  }
-
-  loadFiles (e) {
-    // Read files as data URL and call change handlers
-    const files = [...e.target.files] // when multiple=false, `files` is still array-like
     
-    // Do not reload files that have been successfully loaded
-    const filesToLoad = files.filter((file) => {
-      return !this.state.files.some((existingFile) => existingFile.name === file.name)
-    })
-    
-    return filesToLoad.map((file) => {
-      return readFile(file)
-        .then((fileData) => {
-          this.onChange(fileData, file)
-        })
-    })
+    this.fileInput = null
+    this.setFileInputRef = element => {
+      this.fileInput = element
+    }
+    this.clearFileInput = () => {
+      if (this.fileInput) this.fileInput.value = ""
+    }
   }
   
-  // TODO: Should this fire once or per file?
-  onChange (fileData, file) {
-    // Call redux forms onChange and onLoad callback
-    const { input: { onChange, value }, onLoad, multiple } = this.props
+  async onChange (fileValues) {
+    const { input: { onChange, value: existingFiles }, onRead, multiple } = this.props
     
-    if (multiple) {
-      onChange([...value, fileData])
-      this.setState((state) => {
-        return { files: [ ...state.files, file ] }
-      })
-    } else {
-      onChange(fileData)
-      this.setState({ files: [file] })
+    try {
+      // Only add value to form if file(s) successfully load
+      const filesToAdd = await Promise.resolve(onRead(fileValues)) // wrap in a promise (just in case)
```

Comment:
Since `onRead` now corresponds directly with `onChange`, we can probably accomplish the same effect by just using `onChange`. So CloudinaryInput would look like:
```jsx
...
return (
 <FileInput
      input={{ 
        ...input, 
        onChange: async (files) => {
           // upload, then call input.onChange with the result.
        }
}
```
This should reduce the API surface area and make the return value clearer.

Reply:
So then the onChange in the file input would only call this?

Reply:
Yes, it would have no knowledge of `onChange` being overridden

Reply:
What's your recommendation for handling the remove use case? That's using the input's `onChange` method as well

Reply:
```jsx
<FileInput
  input={{...input, onChange: async (files) => { input.onChange(files)}
  onRemove={ (files) => input.onChange(files) }
/>
```
`onRemove` then is a prop defaults to `this.removeFile` but can be completely overridden?

Reply:
Hmm... so I guess instead of overriding onChange this is really adding an extra read step for new files. That makes sense. Maybe a different name would alleviate confusion? I assumed `onRead` was a callback for triggering side effects but it's really part of the main flow... maybe `postProcess` or something? I'm probably overthinking it 😉Another option could be to overwrite `readFile` and pass that in:

```jsx
...
return (
 <FileInput
      input={ input }
      // called with all new files
      readFiles={ async (files) => {
          // same util used in fileinput
          const filesWithUrls = await readFileUrls(files)
         // upload them to cloudinary and return the result
          
      }}
```

Reply:
I like the latter approach!